tensorizer/serialization.py

```diff
@@ -2432,384 +2181,283 @@ def _generate_state_dict(self) -> None:
-                    raise
+                if not is_cuda:
+                    # Not in CUDA, no pinned memory. Allocate a new buffer for each tensor because that's what we're going to be using long-term
+                    buffer_tensor = torch.empty((needed_buffer_size,), dtype=torch.uint8)
+                    buffer_ptr = buffer_tensor.data_ptr()
 
-        transfer_thread = threading.Thread(
-            target=transfer, name="TensorizerTransfer", daemon=True
-        )
-        read_thread = threading.Thread(target=read_into_buffers, daemon=True)
+                if is_meta:
+                    # meta tensor. Ensure that at least `deserialized_length` bytes are zero
+                    buffer_tensor.zero_()
 
-        with self._optimize_plaid_mode_buffers(keys):
-            transfer_thread.start()
-            read_thread.start()
+                assert buffer_ptr is not None
+                mv = memoryview(ctypes.cast(buffer_ptr, ctypes.POINTER(ctypes.c_byte * tensor_sizes_by_name[header.name])).contents)
+
+                if not is_meta:
+                    start = time.perf_counter()
+
+                    if unsafe_self._encrypted and mv.nbytes > 0:
+                        TensorDeserializer._stream_decrypt(
+                            file_,
+                            unsafe_self._decryption_pool, # decryption_pool safe to be shared
+                            encryption_method,
+                            key,
+                            mv)
+                    else:
+                        file_.readinto(mv)
+
+                    if verify_hash:
+                        unsafe_self._verify_hashes(header.name, header.hashes, header_hashes, mv)
+
+                    duration = time.perf_counter() - start
+                    perf_stats.file_readinto_millisecs.add(int(1000.0 * duration))
+                    perf_stats.file_readinto_bytes.add(mv.nbytes)
+
+                # create a tensor around it and maybe torch.to('cuda')
+                numpy_tensor = _NumpyTensor.from_buffer(
+                    numpy_dtype,
+                    torch_dtype,
+                    header.shape,
+                    mv,
+                )
+                tensor = numpy_tensor.to_tensor()
+
+                if not is_cuda:
+                    # if not cuda then we're not copying away from
+                    # buffer_tensor. Instead buffer_tensor memory lives on as
+                    # the actual tensor. So build a view based of buffer_tensor
+                    # to retain the reference
+                    tensor = buffer_tensor.view(tensor.dtype).view(header.shape)
+
+                stream_context = torch.cuda.stream(cuda_stream) if is_cuda else contextlib.nullcontext()
+                with stream_context:
+                    parameter = unsafe_self._to_torch_parameter(tensor)
+                    if cuda_stream is not None:
+                        cuda_stream.synchronize()
```

Comment:
Yeah this is a good suggestion. May be a relatively significant change of the code.

Reply:
As far as organization, host-to-device `cudaMemcpyAsync` calls cannot actually run concurrently with one another (on the same device), hence having a single thread perform them all separately from the reader, as its own stage, in my original implementation, so something like that is an option. Other than that, we could spawn a CUDA event here in place of `cuda_stream.synchronize()` and wait on it either 1) the moment the memory range is needed again or 2) when this loop is finished, yield the pending parameter(s) into the out queue, and continue reading or returning

Since the improved read times in this PR make up for this optimization not being here, it's also fine to delay that one optimization to another PR, IMO.

Reply:
Yeah I was thinking we use a cuda event for the tensor that is at the start of the buffer, then opportunistically use consecutive free areas of the buffer for future tensors for as long as we can, then wait on that event when we need to start over again.

Ok, yeah, let's do it in a future PR. I'm potentially changing the buffer allocation method anyway (away from torch.empty) so those two things can go pretty well hand-in-hand.

Reply:
You'd still need an event for each tensor with a spot in the buffer because waiting on the first one to finish won't imply that the later ones sharing the buffer have finished. And yeah, a future PR sounds good with the buffer changes.

Reply:
Oh, I presumed cudaMemcpyAsync is at least serialized within its stream. No?

Reply:
marked as future work!